finetrainers/trainer.py

```diff
@@ -679,43 +702,56 @@ def train(self) -> None:
                             if "pooled_prompt_embeds" in text_conditions:
                                 text_conditions["pooled_prompt_embeds"].fill_(0)
 
-                    # These weighting schemes use a uniform timestep sampling and instead post-weight the loss
-                    weights = compute_density_for_timestep_sampling(
-                        weighting_scheme=self.args.flow_weighting_scheme,
-                        batch_size=batch_size,
-                        logit_mean=self.args.flow_logit_mean,
-                        logit_std=self.args.flow_logit_std,
-                        mode_scale=self.args.flow_mode_scale,
-                    )
-                    indices = (weights * self.scheduler.config.num_train_timesteps).long()
-                    sigmas = scheduler_sigmas[indices]
-                    timesteps = (sigmas * 1000.0).long()
-
-                    noise = torch.randn(
-                        latent_conditions["latents"].shape,
-                        generator=generator,
-                        device=accelerator.device,
-                        dtype=weight_dtype,
-                    )
-                    sigmas = expand_tensor_to_dims(sigmas, ndim=latent_conditions["latents"].ndim)
-                    noisy_latents = (1.0 - sigmas) * latent_conditions["latents"] + sigmas * noise
+                    if "calculate_timesteps" in self.model_config.keys():
+                        timesteps = self.model_config["calculate_timesteps"](
+                            scheduler=self.scheduler,
+                            latent_conditions=latent_conditions,
+                            generator=self.state.generator
+                        )
+                    else: # As flow-based calculations are more common for now.
+                        sigmas = self.derive_sigmas_for_flow(batch_size=batch_size, scheduler_sigmas=scheduler_sigmas)
+                        timesteps = (sigmas * 1000.0).long()
+                    
+                    # Same reason as above.
```

Comment:
Because right now, the else branch is more common so it made sense to do it this way. Open to changes.